internal/graph/extensions/live_query/json_patch.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2023 Upbound Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package live_query
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+// Op is a JSON Patch operation.
+type Op string
+
+const (
+	// OpAdd contains "path" and "value" members.
+	OpAdd Op = "add"
+	// OpRemove contains "path" member.
+	OpRemove Op = "remove"
+	// OpReplace contains "path" and "value" members.
+	OpReplace Op = "replace"
+	// OpMove contains "path" and "from" members.
+	OpMove Op = "move"
+)
+
+// Operation is a single JSON Patch operation.
+type Operation struct {
+	Op    Op          `json:"op"`
+	Path  string      `json:"path"`
+	From  string      `json:"from,omitempty"`
+	Value interface{} `json:"value,omitempty"`
+}
+
+// JSONPatchReporter is a simple custom reporter that records the operations needed to
+// transform one value into another.
+type JSONPatchReporter struct {
+	path  cmp.Path
+	patch []Operation
+}
+
+// PushStep implements cmp.Reporter.
+func (r *JSONPatchReporter) PushStep(ps cmp.PathStep) {
+	r.path = append(r.path, ps)
+}
+
+// Report implements cmp.Reporter.
+// we add any unequal operations to the list of operations.
+func (r *JSONPatchReporter) Report(rs cmp.Result) {
+	if rs.Equal() {
+		return
+	}
+	assert(len(r.path) > 0)
```

Comment:
If this panics here, what's the behavior we see at the system level?

Reply:
Is it possible for a misbehaving client to effectively cause a DoS by sending bad data here? What I'm thinking about is between this and some of the other assert calls, if the process panics each time and for example a client just continued to send the same bad request, the process wouldn't be able to handle other client calls, right?

Reply:
that's a concern. however, the assert is inside the json responses diff reporter. This is only possible due to a bug in `cmp` or the comparator itself. At that point the panic will get logged to stderr and the server will keep running. I'll make sure to add backoff to subscription retries in locutus drone command.